File: test_examples/test_example.swift
// Swift test file for funcfinder

import Foundation

// Simple function
func greet(name: String) -> String {
    return "Hello, \(name)!"
}

// Static function
public static func calculate(a: Int, b: Int) -> Int {
    return a + b
}

// Async function
func fetchData() async throws -> Data {
    let url = URL(string: "https://example.com")!
    let (data, _) = try await URLSession.shared.data(from: url)
    return data
}

// Class with methods
public class Person {
    var name: String
    var age: Int

    init(name: String, age: Int) {
        self.name = name
        self.age = age
    }

    func describe() -> String {
        return "\(name), age \(age)"
    }

    static func create(name: String) -> Person {
        return Person(name: name, age: 0)
    }
}

// Struct
struct Point {
    var x: Double
    var y: Double

    func distance(to other: Point) -> Double {
        let dx = x - other.x
        let dy = y - other.y
        return sqrt(dx*dx + dy*dy)
    }

    mutating func move(dx: Double, dy: Double) {
        x += dx
        y += dy
    }
}

// Enum with methods
enum Direction {
    case north, south, east, west

    func opposite() -> Direction {
        switch self {
        case .north: return .south
        case .south: return .north
        case .east: return .west
        case .west: return .east
        }
    }
}

// Protocol
protocol Drawable {
    func draw()
    func resize(scale: Double)
}

// Extension
extension String {
    func reversed() -> String {
        return String(self.reversed())
    }
}

// Generic function
func swap<T>(_ a: inout T, _ b: inout T) {
    let temp = a
    a = b
    b = temp
}

// Closure as property (should not be detected as function)
let multiply: (Int, Int) -> Int = { a, b in
    return a * b
}

// Guard let else (test for false positive fix)
func safeDivide(_ a: Int, _ b: Int) -> Int? {
    guard b != 0 else {
        return nil
    }
    guard let result = Optional(a / b) else {
        return nil
    }
    return result
}

File: test_examples/test_example.rs
// Rust test file for funcfinder

use std::fmt;

// Simple function
fn greet(name: &str) -> String {
    format!("Hello, {}!", name)
}

// Public function
pub fn add(a: i32, b: i32) -> i32 {
    a + b
}

// Generic function
fn swap<T>(a: &mut T, b: &mut T) {
    std::mem::swap(a, b);
}

// Async function
async fn fetch_data(url: &str) -> Result<String, Box<dyn std::error::Error>> {
    Ok(format!("data from {}", url))
}

// Const function
const fn square(x: i32) -> i32 {
    x * x
}

// Struct
struct Point {
    x: f64,
    y: f64,
}

// Impl block for struct
impl Point {
    fn new(x: f64, y: f64) -> Self {
        Point { x, y }
    }

    fn distance(&self, other: &Point) -> f64 {
        let dx = self.x - other.x;
        let dy = self.y - other.y;
        (dx * dx + dy * dy).sqrt()
    }

    fn move_by(&mut self, dx: f64, dy: f64) {
        self.x += dx;
        self.y += dy;
    }
}

// Tuple struct
struct Color(u8, u8, u8);

impl Color {
    fn rgb(r: u8, g: u8, b: u8) -> Self {
        Color(r, g, b)
    }

    fn to_hex(&self) -> String {
        format!("#{:02x}{:02x}{:02x}", self.0, self.1, self.2)
    }
}

// Enum
enum Direction {
    North,
    South,
    East,
    West,
}

impl Direction {
    fn opposite(&self) -> Direction {
        match self {
            Direction::North => Direction::South,
            Direction::South => Direction::North,
            Direction::East => Direction::West,
            Direction::West => Direction::East,
        }
    }
}

// Enum with data
enum Result<T, E> {
    Ok(T),
    Err(E),
}

// Trait
trait Drawable {
    fn draw(&self);
    fn resize(&mut self, scale: f64);
}

trait Area {
    fn area(&self) -> f64;
}

// Struct implementing trait
struct Circle {
    radius: f64,
}

impl Circle {
    fn new(radius: f64) -> Self {
        Circle { radius }
    }
}

impl Area for Circle {
    fn area(&self) -> f64 {
        std::f64::consts::PI * self.radius * self.radius
    }
}

impl Drawable for Circle {
    fn draw(&self) {
        println!("Drawing circle with radius {}", self.radius);
    }

    fn resize(&mut self, scale: f64) {
        self.radius *= scale;
    }
}

// Generic struct
struct Container<T> {
    value: T,
}

impl<T> Container<T> {
    fn new(value: T) -> Self {
        Container { value }
    }

    fn get(&self) -> &T {
        &self.value
    }

    fn set(&mut self, value: T) {
        self.value = value;
    }
}

// Union (unsafe)
union IntOrFloat {
    i: i32,
    f: f32,
}

// Implementing Display trait
impl fmt::Display for Point {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "({}, {})", self.x, self.y)
    }
}

// Module
mod utils {
    pub fn helper() -> i32 {
        42
    }

    fn private_helper() -> i32 {
        0
    }
}

// Closure (should not be detected as function)
fn main() {
    let multiply = |a, b| a * b;
    let result = multiply(3, 4);
    println!("{}", result);
}
